design/spark-co-exist-with-object-storages.md

```diff
@@ -20,122 +18,178 @@ Make the solution usable by non-lakeFS users by:
 
 ## Proposal: Introducing RouterFileSystem
 
 We would like to implement a [HadoopFileSystem](https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java) 
-that translates object store URIs into lakeFS URIs according to a configurable mapping, and uses a relevant Hadoop file system to 
-perform any file system operation.  
+that translates object store URIs into lakeFS URIs according to a configurable mapping, and uses a relevant (configured) Hadoop file system to 
+perform any file system operation.
+
+For simplicity, this document uses `s3a` as the only URI scheme used inside the Spark application code, but this solution holds for any other URI scheme.  
 
 ### Handle any interaction with the underlying object store
 
-To allow Spark users to integrate with lakeFS without changing their Spark code, `RouterFileSystem` is configured to be 
-the file system for object store URIs with a certain scheme. For example, the following Spark configurations make 
-`RouterFileSystem` the file system for URIs with `scheme=s3`:
-```shell
-fs.s3.impl=RouterFileSystem
+To allow Spark users to integrate with lakeFS without changing their Spark code, we configure `RouterFileSystem` as the 
+file system for the URI scheme (or schemes) their Spark application is using. For example, for a Spark application that 
+reads and writes from S3 using `S3AFileSystem`, we configure `RouterFileSystem` to be the file system for URIs with `scheme=s3a`, as follows: 
+```properties
+fs.s3a.impl=RouterFileSystem
 ```
-This will force any file system operation performed on an ibject URI with `scheme=s3` to go through `RouterFileSystem`.   
+This will force any file system operation performed on an object URI with `scheme=s3a` to go through `RouterFileSystem` before it 
+interacts with the underlying storage.
 
 ### URI translation
 
-`RouterFileSystem` has access to a configurable mapping that maps bucket names into lakeFS repositories and prefixes, e.g.
-```yaml
-MapperConfig = {
-    {
-        "source_bucket_pattern": "bucket-a",
-        "mapped_bucket_name": "example-repo",
-        "mapped_prefix": "dev/"
-    }
-}
+`RouterFileSystem` has access to a configurable mapping that maps any object store URI to any type of URI. This proposal 
+uses s3 object store paths (with URI `scheme=s3a`) and lakeFS paths as examples.
+
+#### Mapping configurations
+
+The mapping configurations are spark properties of the following form:
+```properties
+routerfs.mapping.${mappingIdx}.${toFsScheme}.replace='^${fromFsScheme}://bucket/prefix'
+routerfs.mapping.${mappingIdx}.${toFsScheme}.with='${toFsScheme}://another-bucket/prefix/*'
```

Comment:
You might consider replacing the `${mappingIdx}` and `${toFsScheme}` position (i.e. `${toFsScheme}.${mappingIdx}`) for readability (the index is running "inside" of the scheme)